compiler/rustc_lint/src/context.rs

```diff
@@ -1018,3 +1019,10 @@ impl<'tcx> LayoutOf for LateContext<'tcx> {
         self.tcx.layout_of(self.param_env.and(ty))
     }
 }
+
+pub fn parse_lint_and_tool_name(lint_name: &str) -> (Option<Symbol>, &str) {
+    match lint_name.split_once("::") {
+        Some((tool_name, lint_name)) => (Some(Symbol::intern(tool_name)), lint_name),
+        None => (None, lint_name),
+    }
+}
```

Comment:
The correct way to implement this function is to factor out this code

https://github.com/rust-lang/rust/blob/master/compiler/rustc_lint/src/levels.rs#L324-L348

from `LintLevelsBuilder::push`.

The current implementation doesn't check whether the tool is known or not.

Reply:
It looks like we have two places where we check if the tool is known: the place you mentioned, and also in `check_lint_name` (https://github.com/rust-lang/rust/blob/master/compiler/rustc_lint/src/context.rs#L414-L441). It'd probably be good to combine them, but I think the current implementation does check if the tool is known.

Reply:
>but I think the current implementation does check if the tool is known.

Could you add two test cases for that then?
```rust
// compile-flags: -A unknown_tool::foo <-- Error

fn main() {}
```
```rust
// compile-flags: -A known_tool::foo <-- Ok

#![register_tool(known_tool)]

fn main() {}
```

Reply:
Ahh, I see my confusion. Clippy gets implicitly registered (it looks like [here]()), so the behavior is not to ignore lints for tools that are not defined. Clippy is just special.

The first test case you suggested looks like the same one I added, but it's supposed to fail while I assumed it was supposed to pass.

I'll add the new test cases and update my code to work the way it should.